specs/interpolation.yml

```diff
@@ -197,6 +197,13 @@ tests:
     template: '{{#a}}{{b.c}}{{/a}}'
     expected: ''
 
+  - name: Dotted Names are not valid keys
+    desc: Dotted names shall not be parsed as valid keys
+    data:
+      a.b: c
+    template: '{{a.b}}'
+    expected: ''
```

Comment:
The only reason I caution this approach is that it still does not define what to do in the case that you have a key of the form `a.b`: is that a valid key or not? This is relevant for us because the java and python versions disagree: this is undefined behaviour from the spec.

Reply:
That is, what should the expected output of the following be? At the moment the spec does not define that.
```yaml
data:
  a.b: c
template: '{{a.b}}'
```

Reply:
I think the spec defines it (in the human-oriented text), but does not test for it explicitly. If you follow the human-oriented text to the letter, I agree with you that the output should be the empty string. I also think that the other tests already imply this; they just don't enforce it.

The reason I'm hesitant to add your original version of this test, is that it actively prevents implementations from being more lenient towards their users as an added service. I understand that this is exactly what you are trying to achieve, but I'm not yet convinced that this will make the world a better place. Many implementations have extras outside of the spec, and I think this is a good thing in general. Surely, mustache.java is not harming anyone by allowing a fallback to a property by the name `a.b` in the context?

Reply:
What we did that caused us to run into this was the following: we have a java backend that performs some templating action, and we wanted to provide a python command-line utility that did the same for users to test our templating service. The problem is, this difference means that we cannot do that; moreover, the proposed spec fix would not have resolved this issue for us either.

In a sense, the java implementation follows the letter of the spec, but not the spirit. If it is intended that keys not have periods in them, then why not specify that and enforce it?

It's not the end of the world for us; we can change the data that we use. However, the undefined behaviour in the specification means you can't at the moment have a java and python client produce the same output on the same input. Is it the end of the world? No. Can it cause people issuse? Yes.

Reply:
You have the same problem when one implementation supports lambda arguments (like `{{# someLambda 123 }} ... {{/ someLambda }}`) and the other does not. Is that a reason to explicitly forbid lambda arguments in the spec?

If there is a way to feed the same data and same templates to two implementations and get the result you want in both cases, is it a problem that there exist inputs for which they don't give the same results?

Reply:
I do think there is a difference there though: as I understand lambdas, they are an explicit opt-in feature with tests that describe their behaviour. However, the situation here is one where there is a human-readable description for expected behaviour (periods should not be in keys!), but no tests for that which allows different implementations to do whatever they like, in an undocumented way.

With Lambdas, I presume that the implementations would be clear about this behaviour. With undefined behaviour, I just have to test each one manually to see what they do.

Again, if keys should not have periods, why not have a test for that?

(That said, I also agree that your test with data of type
```yaml
data:
  a.b: c
  a: {b: d}
```
is useful as it is morally a slightly different case)

Reply:
I wrote lambda *arguments*, which do not exist according to the spec (not even according to the optional `~lambdas.yml`). That is, according to the spec, you can write `{{someLambda}}` where `someLambda` is a function or other callable object in the context, but `{{someLambda 123}}` meaning "call function `someLambda` and pass it `123` as an argument` is not a thing. However, some implementations support such notation as an unofficial extension.

Sorry if that was not clear. I hope you can now see that the comparison is quite similar: in both cases, implementations are doing something that does not appear anywhere in the spec, not even in an optional module. My point is that this is not necessarily a reason to disallow it.

Reply:
Fair enough, I take your point that lambda arguments run counter to

> When used as the data value for an Interpolation tag, the lambda MUST be treatable as an arity 0 function, and invoked as such.

in the same way that this runs counter to it. I do think there is a different underlying mechanism though; I doubt that (although @spullara can confirm this) there was an intent to have tags with `.` in them, so much as it was probably not considered. Thus it seems weird to consider it an extension.

Note that _regardless_, adding either the test I wrote or the one that you suggest will break the Java implementation as it currently exists. Why not go for the simpler solution of having keys with `.`in them always be interpreted in the same way by all implementations?

That said, I think ultimately this seems we have a difference of opinion on this. I do think that it is simpler to avoid any possible ambiguity and to have the tests actually match the stated documentation:

>  This tag's content names the data to replace the tag.  A single period (`.`)
>  indicates that the item currently sitting atop the context stack should be
> used; otherwise, name resolution is as follows:
>    1) Split the name on periods; the first part is the name to resolve, any
>    remaining parts should be retained.
>    ...
>    5) If any name parts were retained in step 1, each should be resolved
>    against a context stack containing only the result from the former
>    resolution.  **If any part fails resolution, the result should be considered**
>    **falsey, and should interpolate as the empty string.** (emph. mine)

I think that beyond this, I have to abide by your decision. Let me know if you would like me to add a second test as you describe, or replace my test with yours.

Reply:
> Fair enough, (...). I do think there is a different underlying mechanism though; I doubt that (although @spullara can confirm this) there was an intent to have tags with `.` in them, so much as it was probably not considered. Thus it seems weird to consider it an extension.

I don't think the behavior could happen by accident, unless mustache.java also fails this one:

https://github.com/mustache/spec/blob/83b221aec50adb91eeaa935ba7bd51cd792daece/specs/interpolation.yml#L144-L148

That said, you are right that @spullara should have the last word on this.

> 
> Note that _regardless_, adding either the test I wrote or the one that you suggest will break the Java implementation as it currently exists. Why not go for the simpler solution of having keys with `.`in them always be interpreted in the same way by all implementations?

Because there is still a difference in consequences with regard to compatibility. If an implementation fails the test I suggested, that implementation is necessarily breaking output that other implementations would render correctly. If an implementation fails the test you suggested, it only means it is doing something arguably useful (if sloppy) where other implementations would do nothing useful at all.

> 
> (...) I do think that it is simpler to avoid any possible ambiguity and to have the tests actually match the stated documentation:
> 
> > (...)
> > 
> > 2. If any name parts were retained in step 1, each should be resolved
> >    against a context stack containing only the result from the former
> >    resolution.  **If any part fails resolution, the result should be considered**
> >    **falsey, and should interpolate as the empty string.** (emph. mine)

Right. As I wrote before, I agree your test is already implied by the description and also by the other tests. Still, I do not think it is generally necessary of even desirable to explicitly forbid every possible nonstandard behavior.

> 
> I think that beyond this, I have to abide by your decision.

It is not so simple, but the reality is better.

I can choose to accept or reject your proposal, but the content is up to you. Likewise, you cannot control whether your proposal is merged, but you do have the power to adjust or withdraw your proposal. Ideally, we reach some form of consensus and I end up merging something that you can stand behind. I do think we are already closer to consensus than we were at the start.

Besides that, I share the maintainership with two other people, so it is not up to me alone to decide over approval or rejection. Even if I were to reject, your PR could still get merged if the other two maintainers approve, or vice versa.

In this case, I would like to specifically hear @spullara's opinion, because his implementation is known to be affected by these tests. If we don't hear back from him soon, I will email him.

> Let me know if you would like me to add a second test as you describe, or replace my test with yours.

If you agree, please add my suggestion as a second test for the time being. This makes it a bit easier to compare the options for new reviewers. Later on in the review process, we might remove one of the tests again.

Reply:
I have added your suggested test in a separate commit.

Reply:
I'm happy with the more explicit specification that removes the ambiguity. I think you should have both test cases mentioned in here. Have to figure out how to make it easy for folks to get backwards compatibility. For instance, I can imagine folks passing in an object that has all paths expanded before handing them to mustache.java for performance reasons.

Reply:
@spullara Maybe a customization flag (old/new behavior)?

Reply:
@bobthecow do you agree both tests should go in? Then I'll merge this, after recompiling the JSON.

Reply:
Any news on this? I work with @simon-ess , and we would appreciate some conclusion to this.

Reply:
@anderslindho and @simon-ess, apologies for the long silence. I was waiting for a reply by @bobthecow, but it probably escaped his attention and then it escaped mine as well. However, we can decide without his input because @spullara and I already agree.

Concluding, we will merge this. I can do it in the week of June 17-21.

Reply:
Thank you kindly

Reply:
This escaped my attention again, lots of other things on my plate. I'm writing this down and will let you know when I find another time slot.

Reply:
I'm working on this right now.